### PR TITLE
Striker tactic

### PR DESCRIPTION
**Pull request by Yudai-8-om**
## Description
(Premature) striker tactic (+ assist tactic).

Current function:
Pass -> Receive -> Shoot OR Capture -> Shoot depends on the situation

Targeted point for shoot is selected based on cost function
The striker is selected based on the probability that robot can score the goal

TODO:
need to find better cost func

## Steps to test

1. run sim
2. place ball and robots whatever you like on the opponent's side field
image
something like this
3. Robot that has the highest probability to score the goal receive the ball or capture the ball itself and shoot.

Expected result: Goal!

## Concerns
Behavior is a bit unstable

### CODE REVIEW

File: rj_gameplay/rj_gameplay/skill/shoot.py [hunk 1]
```diff
@@ -10,32 +10,42 @@
...
-import stp.skill.sequence as sequence
+import numpy as np
 
 class IShoot(skill.ISkill, ABC):
     ...
 
 """
 A shoot skill which aims at the goal and shoots
 """
+random_shoot = random.uniform(-0.5,0.5)
```
Comment (kfu02) [inline rj_gameplay/rj_gameplay/skill/shoot.py]:
```suggestion
random_shoot = np.random.uniform(-0.5,0.5)
```

Sim doesn't launch without this

File: rj_gameplay/rj_gameplay/tactic/striker_tactic.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+		return role_requests
+
+	def tick(self, role_results: tactic.RoleResults) -> List[tactic.SkillEntry]:
+		"""
+		:return: list of skills
+		"""
+		capture_result = role_results[self.capture]
+		shoot_result = role_results[self.shoot]
+		if capture_result and capture_result[0].is_filled():
```
Comment (kfu02) [inline rj_gameplay/rj_gameplay/tactic/striker_tactic.py]: I did not know there was an `is_filled()` method. Good to know

File: rj_gameplay/rj_gameplay/tactic/striker_tactic.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+
+class StrikerTactic(tactic.ITactic):
+	"""
+	A striker tactic which captures then shoots the ball
+	"""
+
+	def __init__(self):
+		self.capture = tactic.SkillEntry(capture.Capture()) 
+		#  <- receive skill is not published yet
```
Comment (kfu02) [inline rj_gameplay/rj_gameplay/tactic/striker_tactic.py]:
```suggestion
		#  TODO: <- receive skill is not published yet
```

File: rj_gameplay/stp/skill/action_behavior.py [hunk 4]
```diff
@@ -42,8 +53,7 @@ def update(self) -> py_trees.common.Status:
         Check action and return the current state of the aciton
         TODO: Needs to somehow use robot intents to check on status of action
         """
-
-        if self.action.is_done(self.world_state):
+        if not self.world_state is None and self.action.is_done(self.world_state):
```
Comment (kfu02) [inline rj_gameplay/stp/skill/action_behavior.py]:
```suggestion
        if self.world_state is not None and self.action.is_done(self.world_state):
```

**Review by kfu02 — CHANGES_REQUESTED** [review]
Looks good so far (except one small bug). Looking forward to reviewing again once you have it kicking toward goal, haha

File: rj_gameplay/rj_gameplay/play/striker.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		skills = self.striker_tactic.tick(role_results[self.striker_tactic])
+		skill_dict = {}
+
+		skill_dict.update(role_results[self.striker_tactic])
+
+		return (skill_dict, skills)
+
+	def is_done(self, world_state):
+		return self.striker_tactic.is_done(world_state)
+
```
Comment (HussainGynai) [inline rj_gameplay/rj_gameplay/play/striker.py:54]: I think you have some tabs or spaces on this empty line so its not a new line.

File: rj_gameplay/rj_gameplay/tactic/striker_tactic.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+class StrikerTactic(tactic.ITactic):
+	"""
+	A striker tactic which captures then shoots the ball
+	"""
+
+	def __init__(self):
+		self.capture = tactic.SkillEntry(capture.Capture()) 
+		#TODO: add receive skill 
+		self.cost = shoot_cost()
```
Comment (HussainGynai) [inline rj_gameplay/rj_gameplay/tactic/striker_tactic.py]: Add a param for a cost function in the init, so we can put in any one we want in it.

**Review by HussainGynai — CHANGES_REQUESTED** [review]

**Review by HussainGynai — CHANGES_REQUESTED** [review]
For me, when I test, two robots just pass between each other over and over, or one of the robots gets the ball and pivots randomly back and forth.

**Pull request closed without merging**